category/category.go

```diff
@@ -0,0 +1,262 @@
+
+// Equal returns true if two Category objects are equal; otherwise false is returned.
+func (category Category) Equal(u convert.Equaler) bool {
+	other, ok := u.(Category)
+	if !ok {
+		return false
+	}
+	if !uuid.Equal(category.ID, other.ID) {
+		return false
+	}
+	if !category.Lifecycle.Equal(other.Lifecycle) {
+		return false
+	}
+	if category.Name != other.Name {
+		return false
+	}
+	return true
+}
+
+// Repository encapsulates storage and retrieval of categories
+type Repository interface {
+	Create(ctx context.Context, category *Category) (*Category, error)
+	LoadCategoryFromDB(ctx context.Context, id uuid.UUID) (*Category, error)
+	List(ctx context.Context) ([]*Category, error)
+	CreateRelationship(ctx context.Context, relationship *WorkItemTypeCategoryRelationship) error
+	LoadWorkItemTypeCategoryRelationship(ctx context.Context, workitemtypeID uuid.UUID, categoryID uuid.UUID) (*WorkItemTypeCategoryRelationship, error)
+	LoadAllRelationshipsOfCategory(ctx context.Context, categoryID uuid.UUID) ([]*WorkItemTypeCategoryRelationship, error)
+	Save(ctx context.Context, category *Category) (*Category, error)
+}
+
+// NewRepository creates a new storage type.
+func NewRepository(db *gorm.DB) Repository {
+	return &GormRepository{db: db}
+}
+
+// GormRepository is the implementation of the storage interface for Categories.
+type GormRepository struct {
+	db *gorm.DB
+}
+
+// List all Categories.
+func (m *GormRepository) List(ctx context.Context) ([]*Category, error) {
+	var objs []*Category
+	err := m.db.Find(&objs).Error
+	if err != nil && err != gorm.ErrRecordNotFound {
+		log.Error(ctx, map[string]interface{}{
+			"err": err,
+		}, "unable to list categories")
+		return nil, errs.WithStack(err)
+	}
+	return objs, nil
+}
+
+// CreateRelationship creates relationship between workitemtype and category
+func (m *GormRepository) CreateRelationship(ctx context.Context, relationship *WorkItemTypeCategoryRelationship) error {
+	if relationship.ID == uuid.Nil {
+		relationship.ID = uuid.NewV4()
+	}
+	db := m.db.Create(relationship)
+	if db.Error != nil {
+		if gormsupport.IsUniqueViolation(db.Error, "workitemtype_categories_idx") {
+			return errors.NewBadParameterError("category+workitemtype", relationship.CategoryID).Expected("unique")
+		} else {
+			return errors.NewInternalError(db.Error.Error())
+		}
+	}
+	return nil
+}
+
+// Create creates category. This function is used to populate categories table during migration -> PopulateCategories()
+func (m *GormRepository) Create(ctx context.Context, category *Category) (*Category, error) {
+	if category.ID == uuid.Nil {
+		category.ID = uuid.NewV4()
+	}
+	if category.Name == "" {
+		return nil, errors.NewBadParameterError("Name", category.Name)
+
+	}
+	db := m.db.Create(category)
+	if db.Error != nil {
+		if gormsupport.IsUniqueViolation(db.Error, "categories_name_idx") {
+			return nil, errors.NewBadParameterError("Name", category.Name).Expected("unique")
+		} else {
```

Comment:
golint: `if block ends with a return statement, so drop this else and outdent its block`